grudge/reductions.py

```diff
@@ -152,12 +158,15 @@ def nodal_sum_loc(dcoll: DiscretizationCollection, dd, vec) -> float:
 
     :arg dd: a :class:`~grudge.dof_desc.DOFDesc`, or a value
         convertible to one.
-    :arg vec: a :class:`~meshmode.dof_array.DOFArray`.
+    :arg vec: a :class:`~meshmode.dof_array.DOFArray` or an
+        :class:`~arraycontext.container.ArrayContainer`.
     :returns: a scalar denoting the rank-local nodal sum.
     """
-    if isinstance(vec, np.ndarray):
-        return sum(nodal_sum_loc(dcoll, dd, vec[idx])
-                   for idx in np.ndindex(vec.shape))
+    if not isinstance(vec, DOFArray):
+        return sum(
+            nodal_sum_loc(dcoll, dd, comp)
+            for _, comp in serialize_container(vec)
+        )
```

Comment:
I wonder if these reduction functions (and `norm`) should have an optional flag indicating whether to sum across components or not? I could imagine wanting to take a `norm` of a `ConservedVars` and get back a `ConservedVars` containing the norm of each component.